In [2]:
import numpy as np
import pandas as pd

### Dataframe core components

**Two-Dimensional**: It organizes data into a grid of rows and columns (an $X$ and $Y$ axis).

**Heterogeneous**: Unlike a NumPy array, a single DataFrame can store a string in Column A, an integer in Column B, and a boolean in Column C.

**Labeled Axes**: It uses an Index for rows and Columns for vertical headers, allowing you to access data by name (e.g., df['Price']) rather than just by position.

### Creating dataframes

In [3]:
#using lists

investment_data = [['Eternal', 10, 230, 2300], ['Jio Financial', 10, 230, 2300]]

pd.DataFrame(investment_data, columns=['Stock Name', 'Quantity', 'Price per share', 'Total Amount'])

,Stock Name,Quantity,Price per share,Total Amount
0,Eternal,10,230,2300
1,Jio Financial,10,230,2300


In [5]:
# using dictionary

investment_data_dict = {
    'Stock Name': ['Eternal', 'Jio Financial'],
    'Quantity': [10,10],
    'Price per share': [230,230],
    'Total Amount': [2300,2300]
}

stocks_data = pd.DataFrame(investment_data_dict)

stocks_data

,Stock Name,Quantity,Price per share,Total Amount
0,Eternal,10,230,2300
1,Jio Financial,10,230,2300


In [6]:
# using read_csv

ipl_data = pd.read_csv('ipl-matches.csv')

ipl_data

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
0,1312200,Ahmedabad,2022-05-29,2022,Final,Rajasthan Royals,Gujarat Titans,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,bat,N,Gujarat Titans,Wickets,7.0,NaN,HH Pandya,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",CB Gaffaney,Nitin Menon
1,1312199,Ahmedabad,2022-05-27,2022,Qualifier 2,Royal Challengers Bangalore,Rajasthan Royals,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,field,N,Rajasthan Royals,Wickets,7.0,NaN,JC Buttler,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...",CB Gaffaney,Nitin Menon
2,1312198,Kolkata,2022-05-25,2022,Eliminator,Royal Challengers Bangalore,Lucknow Super Giants,"Eden Gardens, Kolkata",Lucknow Super Giants,field,N,Royal Challengers Bangalore,Runs,14.0,NaN,RM Patidar,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['Q de Kock', 'KL Rahul', 'M Vohra', 'DJ Hooda...",J Madanagopal,MA Gough
3,1312197,Kolkata,2022-05-24,2022,Qualifier 1,Rajasthan Royals,Gujarat Titans,"Eden Gardens, Kolkata",Gujarat Titans,field,N,Gujarat Titans,Wickets,7.0,NaN,DA Miller,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",BNJ Oxenford,VK Sharma
4,1304116,Mumbai,2022-05-22,2022,70,Sunrisers Hyderabad,Punjab Kings,"Wankhede Stadium, Mumbai",Sunrisers Hyderabad,bat,N,Punjab Kings,Wickets,5.0,NaN,Harpreet Brar,"['PK Garg', 'Abhishek Sharma', 'RA Tripathi', ...","['JM Bairstow', 'S Dhawan', 'M Shahrukh Khan',...",AK Chaudhary,NA Patwardhan
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
945,335986,Kolkata,2008-04-20,2007/08,4,Kolkata Knight Riders,Deccan Chargers,Eden Gardens,Deccan Chargers,bat,N,Kolkata Knight Riders,Wickets,5.0,NaN,DJ Hussey,"['WP Saha', 'BB McCullum', 'RT Ponting', 'SC G...","['AC Gilchrist', 'Y Venugopal Rao', 'VVS Laxma...",BF Bowden,K Hariharan
946,335985,Mumbai,2008-04-20,2007/08,5,Mumbai Indians,Royal Challengers Bangalore,Wankhede Stadium,Mumbai Indians,bat,N,Royal Challengers Bangalore,Wickets,5.0,NaN,MV Boucher,"['L Ronchi', 'ST Jayasuriya', 'DJ Thornely', '...","['S Chanderpaul', 'R Dravid', 'LRPL Taylor', '...",SJ Davis,DJ Harper
947,335984,Delhi,2008-04-19,2007/08,3,Delhi Daredevils,Rajasthan Royals,Feroz Shah Kotla,Rajasthan Royals,bat,N,Delhi Daredevils,Wickets,9.0,NaN,MF Maharoof,"['G Gambhir', 'V Sehwag', 'S Dhawan', 'MK Tiwa...","['T Kohli', 'YK Pathan', 'SR Watson', 'M Kaif'...",Aleem Dar,GA Pratapkumar
948,335983,Chandigarh,2008-04-19,2007/08,2,Kings XI Punjab,Chennai Super Kings,"Punjab Cricket Association Stadium, Mohali",Chennai Super Kings,bat,N,Chennai Super Kings,Runs,33.0,NaN,MEK Hussey,"['K Goel', 'JR Hopes', 'KC Sangakkara', 'Yuvra...","['PA Patel', 'ML Hayden', 'MEK Hussey', 'MS Dh...",MR Benson,SL Shastri


In [7]:
movies_data = pd.read_csv('movies.csv')
movies_data

,title_x,imdb_id,poster_path,wiki_link,title_y,original_title,is_adult,year_of_release,runtime,genres,imdb_rating,imdb_votes,story,summary,tagline,actors,wins_nominations,release_date
0,Uri: The Surgical Strike,tt8291224,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Uri:_The_Surgica...,Uri: The Surgical Strike,Uri: The Surgical Strike,0,2019,138,Action|Drama|War,8.4,35112,Divided over five chapters the film chronicle...,Indian army special forces execute a covert op...,NaN,Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...,4 wins,11 January 2019 (USA)
1,Battalion 609,tt9472208,NaN,https://en.wikipedia.org/wiki/Battalion_609,Battalion 609,Battalion 609,0,2019,131,War,4.1,73,The story revolves around a cricket match betw...,The story of Battalion 609 revolves around a c...,NaN,Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...,NaN,11 January 2019 (India)
2,The Accidental Prime Minister (film),tt6986710,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/The_Accidental_P...,The Accidental Prime Minister,The Accidental Prime Minister,0,2019,112,Biography|Drama,6.1,5549,Based on the memoir by Indian policy analyst S...,Explores Manmohan Singh's tenure as the Prime ...,NaN,Anupam Kher|Akshaye Khanna|Aahana Kumra|Atul S...,NaN,11 January 2019 (USA)
3,Why Cheat India,tt8108208,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Why_Cheat_India,Why Cheat India,Why Cheat India,0,2019,121,Crime|Drama,6.0,1891,The movie focuses on existing malpractices in ...,The movie focuses on existing malpractices in ...,NaN,Emraan Hashmi|Shreya Dhanwanthary|Snighdadeep ...,NaN,18 January 2019 (USA)
4,Evening Shadows,tt6028796,NaN,https://en.wikipedia.org/wiki/Evening_Shadows,Evening Shadows,Evening Shadows,0,2018,102,Drama,7.3,280,While gay rights and marriage equality has bee...,Under the 'Evening Shadows' truth often plays...,NaN,Mona Ambegaonkar|Ananth Narayan Mahadevan|Deva...,17 wins & 1 nomination,11 January 2019 (India)
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1624,Tera Mera Saath Rahen,tt0301250,https://upload.wikimedia.org/wikipedia/en/2/2b...,https://en.wikipedia.org/wiki/Tera_Mera_Saath_...,Tera Mera Saath Rahen,Tera Mera Saath Rahen,0,2001,148,Drama,4.9,278,Raj Dixit lives with his younger brother Rahu...,A man is torn between his handicapped brother ...,NaN,Ajay Devgn|Sonali Bendre|Namrata Shirodkar|Pre...,NaN,7 November 2001 (India)
1625,Yeh Zindagi Ka Safar,tt0298607,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Yeh_Zindagi_Ka_S...,Yeh Zindagi Ka Safar,Yeh Zindagi Ka Safar,0,2001,146,Drama,3.0,133,Hindi pop-star Sarina Devan lives a wealthy ...,A singer finds out she was adopted when the ed...,NaN,Ameesha Patel|Jimmy Sheirgill|Nafisa Ali|Gulsh...,NaN,16 November 2001 (India)
1626,Sabse Bada Sukh,tt0069204,NaN,https://en.wikipedia.org/wiki/Sabse_Bada_Sukh,Sabse Bada Sukh,Sabse Bada Sukh,0,2018,\N,Comedy|Drama,6.1,13,Village born Lalloo re-locates to Bombay and ...,Village born Lalloo re-locates to Bombay and ...,NaN,Vijay Arora|Asrani|Rajni Bala|Kumud Damle|Utpa...,NaN,NaN
1627,Daaka,tt10833860,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Daaka,Daaka,Daaka,0,2019,136,Action,7.4,38,Shinda tries robbing a bank so he can be wealt...,Shinda tries robbing a bank so he can be wealt...,NaN,Gippy Grewal|Zareen Khan|,NaN,1 November 2019 (USA)


### DataFrame attributes and methods

In [8]:
#shape

ipl_data.shape

(950, 20)

In [10]:
#dtypes

ipl_data.dtypes

ID                   int64
City                object
Date                object
Season              object
MatchNumber         object
Team1               object
Team2               object
Venue               object
TossWinner          object
TossDecision        object
SuperOver           object
WinningTeam         object
WonBy               object
Margin             float64
method              object
Player_of_Match     object
Team1Players        object
Team2Players        object
Umpire1             object
Umpire2             object
dtype: object

In [11]:
#index

ipl_data.index

RangeIndex(start=0, stop=950, step=1)

In [12]:
movies_data.index

RangeIndex(start=0, stop=1629, step=1)

In [13]:
#columns

ipl_data.columns

Index(['ID', 'City', 'Date', 'Season', 'MatchNumber', 'Team1', 'Team2',
       'Venue', 'TossWinner', 'TossDecision', 'SuperOver', 'WinningTeam',
       'WonBy', 'Margin', 'method', 'Player_of_Match', 'Team1Players',
       'Team2Players', 'Umpire1', 'Umpire2'],
      dtype='object')

In [16]:
movies_data.columns

Index(['title_x', 'imdb_id', 'poster_path', 'wiki_link', 'title_y',
       'original_title', 'is_adult', 'year_of_release', 'runtime', 'genres',
       'imdb_rating', 'imdb_votes', 'story', 'summary', 'tagline', 'actors',
       'wins_nominations', 'release_date'],
      dtype='object')

In [17]:
#values

ipl_data.values

array([[1312200, 'Ahmedabad', '2022-05-29', ...,
        "['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pandya', 'DA Miller', 'R Tewatia', 'Rashid Khan', 'R Sai Kishore', 'LH Ferguson', 'Yash Dayal', 'Mohammed Shami']",
        'CB Gaffaney', 'Nitin Menon'],
       [1312199, 'Ahmedabad', '2022-05-27', ...,
        "['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D Padikkal', 'SO Hetmyer', 'R Parag', 'R Ashwin', 'TA Boult', 'YS Chahal', 'M Prasidh Krishna', 'OC McCoy']",
        'CB Gaffaney', 'Nitin Menon'],
       [1312198, 'Kolkata', '2022-05-25', ...,
        "['Q de Kock', 'KL Rahul', 'M Vohra', 'DJ Hooda', 'MP Stoinis', 'E Lewis', 'KH Pandya', 'PVD Chameera', 'Mohsin Khan', 'Avesh Khan', 'Ravi Bishnoi']",
        'J Madanagopal', 'MA Gough'],
       ...,
       [335984, 'Delhi', '2008-04-19', ...,
        "['T Kohli', 'YK Pathan', 'SR Watson', 'M Kaif', 'DS Lehmann', 'RA Jadeja', 'M Rawat', 'D Salunkhe', 'SK Warne', 'SK Trivedi', 'MM Patel']",
        'Aleem Dar', 'GA Pratapkumar'],
    

In [20]:
movies_data.values

array([['Uri: The Surgical Strike', 'tt8291224',
        'https://upload.wikimedia.org/wikipedia/en/thumb/3/3b/URI_-_New_poster.jpg/220px-URI_-_New_poster.jpg',
        ...,
        'Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Gautam|Kirti Kulhari|Rajit Kapoor|Ivan Rodrigues|Manasi Parekh|Swaroop Sampat|Riva Arora|Yogesh Soman|Fareed Ahmed|Akashdeep Arora|Kallol Banerjee|',
        '4 wins', '11 January 2019 (USA)'],
       ['Battalion 609', 'tt9472208', nan, ...,
        'Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elena Kazan|Vishwas Kini|Major Kishore|Jashn Kohli|Rammy C. Pandey|Manish Sharma|Sparsh Sharma|Farnaz Shetty|Vikas Shrivastav|Chandraprakash Thakur|Brajesh Tiwari|',
        nan, '11 January 2019 (India)'],
       ['The Accidental Prime Minister (film)', 'tt6986710',
        'https://upload.wikimedia.org/wikipedia/en/thumb/a/a1/The_Accidental_Prime_Minister_film.jpg/220px-The_Accidental_Prime_Minister_film.jpg',
        ...,
        'Anupam Kher|Akshaye Khanna|Aahana Kumra|At

In [22]:
#head()

ipl_data.head()

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
0,1312200,Ahmedabad,2022-05-29,2022,Final,Rajasthan Royals,Gujarat Titans,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,bat,N,Gujarat Titans,Wickets,7.0,NaN,HH Pandya,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",CB Gaffaney,Nitin Menon
1,1312199,Ahmedabad,2022-05-27,2022,Qualifier 2,Royal Challengers Bangalore,Rajasthan Royals,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,field,N,Rajasthan Royals,Wickets,7.0,NaN,JC Buttler,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...",CB Gaffaney,Nitin Menon
2,1312198,Kolkata,2022-05-25,2022,Eliminator,Royal Challengers Bangalore,Lucknow Super Giants,"Eden Gardens, Kolkata",Lucknow Super Giants,field,N,Royal Challengers Bangalore,Runs,14.0,NaN,RM Patidar,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['Q de Kock', 'KL Rahul', 'M Vohra', 'DJ Hooda...",J Madanagopal,MA Gough
3,1312197,Kolkata,2022-05-24,2022,Qualifier 1,Rajasthan Royals,Gujarat Titans,"Eden Gardens, Kolkata",Gujarat Titans,field,N,Gujarat Titans,Wickets,7.0,NaN,DA Miller,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",BNJ Oxenford,VK Sharma
4,1304116,Mumbai,2022-05-22,2022,70,Sunrisers Hyderabad,Punjab Kings,"Wankhede Stadium, Mumbai",Sunrisers Hyderabad,bat,N,Punjab Kings,Wickets,5.0,NaN,Harpreet Brar,"['PK Garg', 'Abhishek Sharma', 'RA Tripathi', ...","['JM Bairstow', 'S Dhawan', 'M Shahrukh Khan',...",AK Chaudhary,NA Patwardhan


In [23]:
#tail()

ipl_data.tail()

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
945,335986,Kolkata,2008-04-20,2007/08,4,Kolkata Knight Riders,Deccan Chargers,Eden Gardens,Deccan Chargers,bat,N,Kolkata Knight Riders,Wickets,5.0,NaN,DJ Hussey,"['WP Saha', 'BB McCullum', 'RT Ponting', 'SC G...","['AC Gilchrist', 'Y Venugopal Rao', 'VVS Laxma...",BF Bowden,K Hariharan
946,335985,Mumbai,2008-04-20,2007/08,5,Mumbai Indians,Royal Challengers Bangalore,Wankhede Stadium,Mumbai Indians,bat,N,Royal Challengers Bangalore,Wickets,5.0,NaN,MV Boucher,"['L Ronchi', 'ST Jayasuriya', 'DJ Thornely', '...","['S Chanderpaul', 'R Dravid', 'LRPL Taylor', '...",SJ Davis,DJ Harper
947,335984,Delhi,2008-04-19,2007/08,3,Delhi Daredevils,Rajasthan Royals,Feroz Shah Kotla,Rajasthan Royals,bat,N,Delhi Daredevils,Wickets,9.0,NaN,MF Maharoof,"['G Gambhir', 'V Sehwag', 'S Dhawan', 'MK Tiwa...","['T Kohli', 'YK Pathan', 'SR Watson', 'M Kaif'...",Aleem Dar,GA Pratapkumar
948,335983,Chandigarh,2008-04-19,2007/08,2,Kings XI Punjab,Chennai Super Kings,"Punjab Cricket Association Stadium, Mohali",Chennai Super Kings,bat,N,Chennai Super Kings,Runs,33.0,NaN,MEK Hussey,"['K Goel', 'JR Hopes', 'KC Sangakkara', 'Yuvra...","['PA Patel', 'ML Hayden', 'MEK Hussey', 'MS Dh...",MR Benson,SL Shastri
949,335982,Bangalore,2008-04-18,2007/08,1,Royal Challengers Bangalore,Kolkata Knight Riders,M Chinnaswamy Stadium,Royal Challengers Bangalore,field,N,Kolkata Knight Riders,Runs,140.0,NaN,BB McCullum,"['R Dravid', 'W Jaffer', 'V Kohli', 'JH Kallis...","['SC Ganguly', 'BB McCullum', 'RT Ponting', 'D...",Asad Rauf,RE Koertzen


In [24]:
#sample()
movies_data.sample(10)

,title_x,imdb_id,poster_path,wiki_link,title_y,original_title,is_adult,year_of_release,runtime,genres,imdb_rating,imdb_votes,story,summary,tagline,actors,wins_nominations,release_date
458,Jai Ho (film),tt1188982,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Jai_Ho_(film),Jai Ho,Jai Ho,0,2014,135,Action|Drama,5.2,15868,An ex-army officer initiates a unique idea of ...,An ex-army officer initiates a unique idea of ...,People's Man,Salman Khan|Daisy Shah|Tabu|Danny Denzongpa|Na...,1 win & 5 nominations,24 January 2014 (USA)
970,De Dana Dan,tt1255951,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/De_Dana_Dan,De Dana Dan,De Dana Dan,0,2009,162,Comedy|Crime,5.3,6357,Singapore-based Nitin Bankar's education was p...,Two down on their luck men tries to kidnap a w...,The great entertainers are back,Akshay Kumar|Sunil Shetty|Katrina Kaif|Sameera...,1 nomination,27 November 2009 (USA)
273,Tere Bin Laden: Dead or Alive,tt5313980,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Tere_Bin_Laden:_...,Tere Bin Laden: Dead Or Alive,Tere Bin Laden: Dead Or Alive,0,2016,104,Comedy,4.3,771,A common man who happens to look like Osama Bi...,A Bollywood director is hired by an American t...,A million-dollar fake,Parthaa Akerkar|Krishna Bansal|Vikrant Chaturv...,NaN,26 February 2016 (USA)
1566,Yeh Kya Ho Raha Hai?,tt0338566,https://upload.wikimedia.org/wikipedia/en/5/59...,https://en.wikipedia.org/wiki/Yeh_Kya_Ho_Raha_...,Yeh Kya Ho Raha Hai?,Yeh Kya Ho Raha Hai?,0,2002,\N,Comedy|Romance,4.8,217,Four young men enter college in order to study...,Four young men enter college in order to study...,NaN,Prashant Chianani|Aamir Ali Malik|Vaibhav Jhal...,NaN,11 October 2002 (India)
265,Sanam Re,tt4262516,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Sanam_Re,Sanam Re,Sanam Re,0,2016,120,Drama|Romance,3.1,2622,A man focused on his career finds solace when ...,A man focused on his career finds solace when ...,Never together... never apart.,Rishi Kapoor|Pulkit Samrat|Yami Gautam|Urvashi...,1 nomination,12 February 2016 (India)
257,Sallu Ki Shaadi,tt7722484,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Sallu_ki_Shaadi,Sallu Ki Shaadi,Sallu Ki Shaadi,0,2017,109,Romance,3.9,28,Zeenat and Shahnawaz were a match made in movi...,Zeenat and Shahnawaz were a match made in movi...,NaN,Zeenat Aman|Sandeep Anand|Asrani|Kashyap Barbh...,NaN,8 December 2017 (India)
963,Bal Ganesh 2,tt8823390,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Bal_Ganesh_2,Bal Ganesh 2,Bal Ganesh 2,0,2009,74,Animation,7.4,10,Lord Ganesha evolves from a mischievous boy w...,Lord Ganesha evolves from a mischievous boy w...,NaN,NaN,NaN,2009 (India)
644,Prague (2013 film),tt2401719,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Prague_(2013_film),Prague,Prague,0,2013,109,Mystery|Romance|Thriller,5.9,313,A city with a history of heritage myth and su...,A psychological thriller about an upcoming arc...,'Men Live in Cities..Cities Live in Men too...',Chandan Roy Sanyal|Arfi Lamba|Kumar Mayank|Son...,NaN,27 September 2013 (India)
870,Once Upon a Time in Mumbaai,tt1395054,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Once_Upon_a_Time...,Once Upon a Time in Mumbai,Once Upon a Time in Mumbaai,0,2010,134,Action|Crime|Drama,7.4,15169,During March 1993 after his Ambassador car is...,A smuggler rises to power in in 1970s Mumbai ...,The man Mumbai once feared. The man Mumbai now...,Ajay Devgn|Emraan Hashmi|Kangana Ranaut|Prachi...,7 wins & 33 nominations,30 July 2010 (India)
392,Hunterrr,tt4333674,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Hunterrr,Hunterrr,Hunterrr,0,2015,141,Comedy|Romance,7.0,5196,Mandar Ponkshe is an ordinary guy meaning the...,Mandar an ordinary guy is obsessed with sex a...,NaN,Gulshan Devaiah

In [25]:
#info()

ipl_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 950 entries, 0 to 949
Data columns (total 20 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   ID               950 non-null    int64  
 1   City             899 non-null    object 
 2   Date             950 non-null    object 
 3   Season           950 non-null    object 
 4   MatchNumber      950 non-null    object 
 5   Team1            950 non-null    object 
 6   Team2            950 non-null    object 
 7   Venue            950 non-null    object 
 8   TossWinner       950 non-null    object 
 9   TossDecision     950 non-null    object 
 10  SuperOver        946 non-null    object 
 11  WinningTeam      946 non-null    object 
 12  WonBy            950 non-null    object 
 13  Margin           932 non-null    float64
 14  method           19 non-null     object 
 15  Player_of_Match  946 non-null    object 
 16  Team1Players     950 non-null    object 
 17  Team2Players    

In [26]:
movies_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1629 entries, 0 to 1628
Data columns (total 18 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   title_x           1629 non-null   object 
 1   imdb_id           1629 non-null   object 
 2   poster_path       1526 non-null   object 
 3   wiki_link         1629 non-null   object 
 4   title_y           1629 non-null   object 
 5   original_title    1629 non-null   object 
 6   is_adult          1629 non-null   int64  
 7   year_of_release   1629 non-null   int64  
 8   runtime           1629 non-null   object 
 9   genres            1629 non-null   object 
 10  imdb_rating       1629 non-null   float64
 11  imdb_votes        1629 non-null   int64  
 12  story             1609 non-null   object 
 13  summary           1629 non-null   object 
 14  tagline           557 non-null    object 
 15  actors            1624 non-null   object 
 16  wins_nominations  707 non-null    object 


In [28]:
#describe(), works only on numerical columns

ipl_data.describe()

,ID,Margin
count,9.500000e+02,932.000000
mean,8.304852e+05,17.056867
std,3.375678e+05,21.633109
min,3.359820e+05,1.000000
25%,5.012612e+05,6.000000
50%,8.297380e+05,8.000000
75%,1.175372e+06,19.000000
max,1.312200e+06,146.000000


In [29]:
movies_data.describe()

,is_adult,year_of_release,imdb_rating,imdb_votes
count,1629.0,1629.000000,1629.000000,1629.000000
mean,0.0,2010.263966,5.557459,5384.263352
std,0.0,5.381542,1.567609,14552.103231
min,0.0,2001.000000,0.000000,0.000000
25%,0.0,2005.000000,4.400000,233.000000
50%,0.0,2011.000000,5.600000,1000.000000
75%,0.0,2015.000000,6.800000,4287.000000
max,0.0,2019.000000,9.400000,310481.000000


In [31]:
#isnull()

ipl_data.isnull() # gives boolean dataframe

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
0,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
945,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False
946,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False
947,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False
948,False,False,False,False,False,False,False,False,False,False,False,False,False,False,True,False,False,False,False,False


In [34]:
ipl_data.isnull().sum() # count of null values column-wise

ID                   0
City                51
Date                 0
Season               0
MatchNumber          0
Team1                0
Team2                0
Venue                0
TossWinner           0
TossDecision         0
SuperOver            4
WinningTeam          4
WonBy                0
Margin              18
method             931
Player_of_Match      4
Team1Players         0
Team2Players         0
Umpire1              0
Umpire2              0
dtype: int64

In [38]:
criteria = ipl_data.isnull().sum() > 20 #filter
criteria.index 

Index(['ID', 'City', 'Date', 'Season', 'MatchNumber', 'Team1', 'Team2',
       'Venue', 'TossWinner', 'TossDecision', 'SuperOver', 'WinningTeam',
       'WonBy', 'Margin', 'method', 'Player_of_Match', 'Team1Players',
       'Team2Players', 'Umpire1', 'Umpire2'],
      dtype='object')

In [39]:
# duplicated, returns true, if found any row twice, otherwise false

movies_data.duplicated()

0       False
1       False
2       False
3       False
4       False
        ...  
1624    False
1625    False
1626    False
1627    False
1628    False
Length: 1629, dtype: bool

In [40]:
# to get the sum

movies_data.duplicated().sum()

np.int64(0)

0 means, there are unique rows.

In [41]:
ipl_data.duplicated()

0      False
1      False
2      False
3      False
4      False
       ...  
945    False
946    False
947    False
948    False
949    False
Length: 950, dtype: bool

In [42]:
ipl_data.duplicated().sum()

np.int64(0)

### .rename()

The .rename() method is used to change the labels of columns or the index in a DataFrame or Series. 

It is highly flexible because you can rename specific labels without having to redefine the entire list of headers.

1. **Renaming Columns**

To change column names, you pass a dictionary to the columns parameter where the key is the old name and the value is the new name.
- **Syntax**: 
            
        df.rename(columns={'old_name': 'new_name'})

- **Output**: 

    A new DataFrame with the updated headers (unless inplace=True is used).
    
- **Common In**: 

    Cleaning up inconsistently named data, removing spaces from headers, or preparing data for a SQL export.
    
2. **Renaming the Index**

Similarly, you can change specific row labels by passing a dictionary to the index parameter.

- **Syntax**: 

        df.rename(index={0: 'Match_1', 1: 'Match_2'})
        
- **Output**: 

A new DataFrame with the updated row labels.

- **Common In**: 

Changing date formats in an index or mapping numeric IDs to descriptive names.

3. **Using Functions (Mappers)**

Instead of a dictionary, you can pass a function (like str.upper or str.lower) to transform all labels at once.

- **Syntax**: 

        df.rename(columns=str.upper)
        
- **Output**: A DataFrame where every column name has been modified by the function.

- **Common In**: Standardizing headers to all lowercase or uppercase for easier coding.

In [45]:
# rename()

movies_data.head(2)

,title_x,imdb_id,poster_path,wiki_link,title_y,original_title,is_adult,year_of_release,runtime,genres,imdb_rating,imdb_votes,story,summary,tagline,actors,wins_nominations,release_date
0,Uri: The Surgical Strike,tt8291224,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Uri:_The_Surgica...,Uri: The Surgical Strike,Uri: The Surgical Strike,0,2019,138,Action|Drama|War,8.4,35112,Divided over five chapters the film chronicle...,Indian army special forces execute a covert op...,NaN,Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...,4 wins,11 January 2019 (USA)
1,Battalion 609,tt9472208,NaN,https://en.wikipedia.org/wiki/Battalion_609,Battalion 609,Battalion 609,0,2019,131,War,4.1,73,The story revolves around a cricket match betw...,The story of Battalion 609 revolves around a c...,NaN,Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...,NaN,11 January 2019 (India)


In [49]:
movies_data.rename(columns={'year_of_release':'release_year'}).head(2)

,title_x,imdb_id,poster_path,wiki_link,title_y,original_title,is_adult,release_year,runtime,genres,imdb_rating,imdb_votes,story,summary,tagline,actors,wins_nominations,release_date
0,Uri: The Surgical Strike,tt8291224,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Uri:_The_Surgica...,Uri: The Surgical Strike,Uri: The Surgical Strike,0,2019,138,Action|Drama|War,8.4,35112,Divided over five chapters the film chronicle...,Indian army special forces execute a covert op...,NaN,Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...,4 wins,11 January 2019 (USA)
1,Battalion 609,tt9472208,NaN,https://en.wikipedia.org/wiki/Battalion_609,Battalion 609,Battalion 609,0,2019,131,War,4.1,73,The story revolves around a cricket match betw...,The story of Battalion 609 revolves around a c...,NaN,Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...,NaN,11 January 2019 (India)


## Mathematical function

### .sum()
The .sum() method calculates the total of all values in a Series or along a specified axis in a DataFrame. By default, it skips missing values (NaN) and returns the arithmetic sum of the data.

- **Syntax**: 

        df.sum(axis=0, skipna=True, numeric_only=False)

    *changing the axis to 1 (or 'columns') calculates the sum horizontally across each row*

- **Output**: A single number (for a Series) or a Series of sums (for a DataFrame).

- **Common In**: Calculating total revenue, total distance traveled, or aggregating scores for a specific time period.

In [50]:
stocks_data

,Stock Name,Quantity,Price per share,Total Amount
0,Eternal,10,230,2300
1,Jio Financial,10,230,2300


In [53]:
stocks_data.sum() # row-sum, Quantiity[0]+Quantity[1]

Stock Name         EternalJio Financial
Quantity                             20
Price per share                     460
Total Amount                       4600
dtype: object

In [59]:
stocks_data_only_numeric = stocks_data.drop(labels=['Stock Name'], axis = 1)

stocks_data_only_numeric

,Quantity,Price per share,Total Amount
0,10,230,2300
1,10,230,2300


In [60]:
stocks_data_only_numeric.sum(axis = 1)

0    2540
1    2540
dtype: int64

### .mean() / .median() / .mode() / .std() / .var()

##### axis = 0 -> row and axis = 1 -> column

In [65]:
stocks_data_only_numeric.mean() #axis =0

Quantity             10.0
Price per share     230.0
Total Amount       2300.0
dtype: float64

In [66]:
stocks_data_only_numeric.mean(axis = 1) #axis = 1

0    846.666667
1    846.666667
dtype: float64

In [67]:
stocks_data_only_numeric.median() #axis =0

Quantity             10.0
Price per share     230.0
Total Amount       2300.0
dtype: float64

In [68]:
stocks_data_only_numeric.median(axis=1)

0    230.0
1    230.0
dtype: float64

In [69]:
stocks_data_only_numeric.mode()

,Quantity,Price per share,Total Amount
0,10,230,2300


In [70]:
stocks_data_only_numeric.mode(axis=1)

,0,1,2
0,10,230,2300
1,10,230,2300


In [71]:
stocks_data_only_numeric.std()

Quantity           0.0
Price per share    0.0
Total Amount       0.0
dtype: float64

In [72]:
stocks_data_only_numeric.var()

Quantity           0.0
Price per share    0.0
Total Amount       0.0
dtype: float64

In [73]:
stocks_data_only_numeric.min()

Quantity             10
Price per share     230
Total Amount       2300
dtype: int64

In [74]:
stocks_data_only_numeric.max()

Quantity             10
Price per share     230
Total Amount       2300
dtype: int64

### Selecting columns from the dataframe

In [75]:
ipl_data

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
0,1312200,Ahmedabad,2022-05-29,2022,Final,Rajasthan Royals,Gujarat Titans,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,bat,N,Gujarat Titans,Wickets,7.0,NaN,HH Pandya,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",CB Gaffaney,Nitin Menon
1,1312199,Ahmedabad,2022-05-27,2022,Qualifier 2,Royal Challengers Bangalore,Rajasthan Royals,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,field,N,Rajasthan Royals,Wickets,7.0,NaN,JC Buttler,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...",CB Gaffaney,Nitin Menon
2,1312198,Kolkata,2022-05-25,2022,Eliminator,Royal Challengers Bangalore,Lucknow Super Giants,"Eden Gardens, Kolkata",Lucknow Super Giants,field,N,Royal Challengers Bangalore,Runs,14.0,NaN,RM Patidar,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['Q de Kock', 'KL Rahul', 'M Vohra', 'DJ Hooda...",J Madanagopal,MA Gough
3,1312197,Kolkata,2022-05-24,2022,Qualifier 1,Rajasthan Royals,Gujarat Titans,"Eden Gardens, Kolkata",Gujarat Titans,field,N,Gujarat Titans,Wickets,7.0,NaN,DA Miller,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",BNJ Oxenford,VK Sharma
4,1304116,Mumbai,2022-05-22,2022,70,Sunrisers Hyderabad,Punjab Kings,"Wankhede Stadium, Mumbai",Sunrisers Hyderabad,bat,N,Punjab Kings,Wickets,5.0,NaN,Harpreet Brar,"['PK Garg', 'Abhishek Sharma', 'RA Tripathi', ...","['JM Bairstow', 'S Dhawan', 'M Shahrukh Khan',...",AK Chaudhary,NA Patwardhan
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
945,335986,Kolkata,2008-04-20,2007/08,4,Kolkata Knight Riders,Deccan Chargers,Eden Gardens,Deccan Chargers,bat,N,Kolkata Knight Riders,Wickets,5.0,NaN,DJ Hussey,"['WP Saha', 'BB McCullum', 'RT Ponting', 'SC G...","['AC Gilchrist', 'Y Venugopal Rao', 'VVS Laxma...",BF Bowden,K Hariharan
946,335985,Mumbai,2008-04-20,2007/08,5,Mumbai Indians,Royal Challengers Bangalore,Wankhede Stadium,Mumbai Indians,bat,N,Royal Challengers Bangalore,Wickets,5.0,NaN,MV Boucher,"['L Ronchi', 'ST Jayasuriya', 'DJ Thornely', '...","['S Chanderpaul', 'R Dravid', 'LRPL Taylor', '...",SJ Davis,DJ Harper
947,335984,Delhi,2008-04-19,2007/08,3,Delhi Daredevils,Rajasthan Royals,Feroz Shah Kotla,Rajasthan Royals,bat,N,Delhi Daredevils,Wickets,9.0,NaN,MF Maharoof,"['G Gambhir', 'V Sehwag', 'S Dhawan', 'MK Tiwa...","['T Kohli', 'YK Pathan', 'SR Watson', 'M Kaif'...",Aleem Dar,GA Pratapkumar
948,335983,Chandigarh,2008-04-19,2007/08,2,Kings XI Punjab,Chennai Super Kings,"Punjab Cricket Association Stadium, Mohali",Chennai Super Kings,bat,N,Chennai Super Kings,Runs,33.0,NaN,MEK Hussey,"['K Goel', 'JR Hopes', 'KC Sangakkara', 'Yuvra...","['PA Patel', 'ML Hayden', 'MEK Hussey', 'MS Dh...",MR Benson,SL Shastri


In [77]:
movies_data.columns

Index(['title_x', 'imdb_id', 'poster_path', 'wiki_link', 'title_y',
       'original_title', 'is_adult', 'year_of_release', 'runtime', 'genres',
       'imdb_rating', 'imdb_votes', 'story', 'summary', 'tagline', 'actors',
       'wins_nominations', 'release_date'],
      dtype='object')

In [76]:
# one feature

ipl_data['City']

0       Ahmedabad
1       Ahmedabad
2         Kolkata
3         Kolkata
4          Mumbai
          ...    
945       Kolkata
946        Mumbai
947         Delhi
948    Chandigarh
949     Bangalore
Name: City, Length: 950, dtype: object

In [79]:
#multiple feature, done by fancy indexing

movies_data[['imdb_id', 'original_title', 'imdb_rating']] # case-sensitive and order matches exactly with the fancy index

,imdb_id,original_title,imdb_rating
0,tt8291224,Uri: The Surgical Strike,8.4
1,tt9472208,Battalion 609,4.1
2,tt6986710,The Accidental Prime Minister,6.1
3,tt8108208,Why Cheat India,6.0
4,tt6028796,Evening Shadows,7.3
...,...,...,...
1624,tt0301250,Tera Mera Saath Rahen,4.9
1625,tt0298607,Yeh Zindagi Ka Safar,3.0
1626,tt0069204,Sabse Bada Sukh,6.1
1627,tt10833860,Daaka,7.4


### .set_index()

The .set_index() method is used to promote one or more existing columns in a DataFrame to become the Index (the row labels). 

This is a fundamental step in data preparation, especially when you want to move away from the default $0, 1, 2, 3...$ numbering.

- **Syntax**: 
        df.set_index(keys, drop=True, append=False, inplace=False)
        
- **Output**: A new DataFrame where the specified column is now the bolded "label" on the left.

- **Common In**: Converting "Date" columns for time-series analysis or using "ID" columns for easier data alignment.

In [82]:
stock_data_modified = stocks_data.set_index(keys='Stock Name')
stock_data_modified

,Quantity,Price per share,Total Amount
Stock Name,,,
Eternal,10,230,2300
Jio Financial,10,230,2300


### Selecting rows from the dataframe

Done by two methods:
1. **iloc** : searches using index positions
2. **loc** : searches using index labels

In Pandas, .loc (Label-based selection) and .iloc (Integer-based selection) are the primary ways to access data in a DataFrame. While they look like functions, they are actually indexers, which is why they use square brackets [] instead of parentheses ().

### .loc (Label-based)

.loc is used to access a group of rows and columns by labels or a boolean array. If you have set a specific column (like 'Match_ID') as your index, you use those IDs to find data.

- **Syntax**: 

        df.loc[row_label, column_label]

- **Slicing**: 
Unlike standard Python slicing, .loc is inclusive of the stop value.

*df.loc[100:105] will give you rows 100, 101, 102, 103, 104, and 105.*

- **Common In**: Selecting specific categories, filtering by date, or updating values based on a condition.

### .iloc (Integer-based)
.iloc is purely positional. It doesn't care what the labels are; it only cares about where the data sits in the table ($0, 1, 2...$).

- **Syntax**: 

        df.iloc[row_position, column_position]
    
- **Slicing**: This follows standard Python rules: exclusive of the stop value.df.iloc[0:5] gives you positions 0, 1, 2, 3, and 4.
- **Common In**: Taking the "first 10 rows," splitting data into training/testing sets, or selecting columns by their index number (e.g., "give me the last column").

In [83]:
movies_data

,title_x,imdb_id,poster_path,wiki_link,title_y,original_title,is_adult,year_of_release,runtime,genres,imdb_rating,imdb_votes,story,summary,tagline,actors,wins_nominations,release_date
0,Uri: The Surgical Strike,tt8291224,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Uri:_The_Surgica...,Uri: The Surgical Strike,Uri: The Surgical Strike,0,2019,138,Action|Drama|War,8.4,35112,Divided over five chapters the film chronicle...,Indian army special forces execute a covert op...,NaN,Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...,4 wins,11 January 2019 (USA)
1,Battalion 609,tt9472208,NaN,https://en.wikipedia.org/wiki/Battalion_609,Battalion 609,Battalion 609,0,2019,131,War,4.1,73,The story revolves around a cricket match betw...,The story of Battalion 609 revolves around a c...,NaN,Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...,NaN,11 January 2019 (India)
2,The Accidental Prime Minister (film),tt6986710,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/The_Accidental_P...,The Accidental Prime Minister,The Accidental Prime Minister,0,2019,112,Biography|Drama,6.1,5549,Based on the memoir by Indian policy analyst S...,Explores Manmohan Singh's tenure as the Prime ...,NaN,Anupam Kher|Akshaye Khanna|Aahana Kumra|Atul S...,NaN,11 January 2019 (USA)
3,Why Cheat India,tt8108208,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Why_Cheat_India,Why Cheat India,Why Cheat India,0,2019,121,Crime|Drama,6.0,1891,The movie focuses on existing malpractices in ...,The movie focuses on existing malpractices in ...,NaN,Emraan Hashmi|Shreya Dhanwanthary|Snighdadeep ...,NaN,18 January 2019 (USA)
4,Evening Shadows,tt6028796,NaN,https://en.wikipedia.org/wiki/Evening_Shadows,Evening Shadows,Evening Shadows,0,2018,102,Drama,7.3,280,While gay rights and marriage equality has bee...,Under the 'Evening Shadows' truth often plays...,NaN,Mona Ambegaonkar|Ananth Narayan Mahadevan|Deva...,17 wins & 1 nomination,11 January 2019 (India)
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1624,Tera Mera Saath Rahen,tt0301250,https://upload.wikimedia.org/wikipedia/en/2/2b...,https://en.wikipedia.org/wiki/Tera_Mera_Saath_...,Tera Mera Saath Rahen,Tera Mera Saath Rahen,0,2001,148,Drama,4.9,278,Raj Dixit lives with his younger brother Rahu...,A man is torn between his handicapped brother ...,NaN,Ajay Devgn|Sonali Bendre|Namrata Shirodkar|Pre...,NaN,7 November 2001 (India)
1625,Yeh Zindagi Ka Safar,tt0298607,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Yeh_Zindagi_Ka_S...,Yeh Zindagi Ka Safar,Yeh Zindagi Ka Safar,0,2001,146,Drama,3.0,133,Hindi pop-star Sarina Devan lives a wealthy ...,A singer finds out she was adopted when the ed...,NaN,Ameesha Patel|Jimmy Sheirgill|Nafisa Ali|Gulsh...,NaN,16 November 2001 (India)
1626,Sabse Bada Sukh,tt0069204,NaN,https://en.wikipedia.org/wiki/Sabse_Bada_Sukh,Sabse Bada Sukh,Sabse Bada Sukh,0,2018,\N,Comedy|Drama,6.1,13,Village born Lalloo re-locates to Bombay and ...,Village born Lalloo re-locates to Bombay and ...,NaN,Vijay Arora|Asrani|Rajni Bala|Kumud Damle|Utpa...,NaN,NaN
1627,Daaka,tt10833860,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Daaka,Daaka,Daaka,0,2019,136,Action,7.4,38,Shinda tries robbing a bank so he can be wealt...,Shinda tries robbing a bank so he can be wealt...,NaN,Gippy Grewal|Zareen Khan|,NaN,1 November 2019 (USA)


In [85]:
movies_data.index

RangeIndex(start=0, stop=1629, step=1)

In [88]:
movies_data.loc[1:10,['original_title', 'imdb_rating', 'imdb_votes']] # extract rows from index = 1 to 10, and the specified columns

,original_title,imdb_rating,imdb_votes
1,Battalion 609,4.1,73
2,The Accidental Prime Minister,6.1,5549
3,Why Cheat India,6.0,1891
4,Evening Shadows,7.3,280
5,Soni,7.2,1595
6,Fraud Saiyyan,4.2,504
7,Bombairiya,4.3,295
8,Manikarnika: The Queen of Jhansi,6.5,7361
9,Thackeray,5.1,2301
10,Amavas,2.8,235


In [91]:
# extract first 10 rows and all columns except last

movies_data.iloc[0:10,:len(movies_data.columns)-1] # last column is excluded

,title_x,imdb_id,poster_path,wiki_link,title_y,original_title,is_adult,year_of_release,runtime,genres,imdb_rating,imdb_votes,story,summary,tagline,actors,wins_nominations
0,Uri: The Surgical Strike,tt8291224,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Uri:_The_Surgica...,Uri: The Surgical Strike,Uri: The Surgical Strike,0,2019,138,Action|Drama|War,8.4,35112,Divided over five chapters the film chronicle...,Indian army special forces execute a covert op...,NaN,Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...,4 wins
1,Battalion 609,tt9472208,NaN,https://en.wikipedia.org/wiki/Battalion_609,Battalion 609,Battalion 609,0,2019,131,War,4.1,73,The story revolves around a cricket match betw...,The story of Battalion 609 revolves around a c...,NaN,Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...,NaN
2,The Accidental Prime Minister (film),tt6986710,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/The_Accidental_P...,The Accidental Prime Minister,The Accidental Prime Minister,0,2019,112,Biography|Drama,6.1,5549,Based on the memoir by Indian policy analyst S...,Explores Manmohan Singh's tenure as the Prime ...,NaN,Anupam Kher|Akshaye Khanna|Aahana Kumra|Atul S...,NaN
3,Why Cheat India,tt8108208,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Why_Cheat_India,Why Cheat India,Why Cheat India,0,2019,121,Crime|Drama,6.0,1891,The movie focuses on existing malpractices in ...,The movie focuses on existing malpractices in ...,NaN,Emraan Hashmi|Shreya Dhanwanthary|Snighdadeep ...,NaN
4,Evening Shadows,tt6028796,NaN,https://en.wikipedia.org/wiki/Evening_Shadows,Evening Shadows,Evening Shadows,0,2018,102,Drama,7.3,280,While gay rights and marriage equality has bee...,Under the 'Evening Shadows' truth often plays...,NaN,Mona Ambegaonkar|Ananth Narayan Mahadevan|Deva...,17 wins & 1 nomination
5,Soni (film),tt6078866,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Soni_(film),Soni,Soni,0,2018,97,Drama,7.2,1595,Soni a young policewoman in Delhi and her su...,While fighting crimes against women in Delhi ...,NaN,Geetika Vidya Ohlyan|Saloni Batra|Vikas Shukla...,3 wins & 5 nominations
6,Fraud Saiyaan,tt5013008,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Fraud_Saiyaan,Fraud Saiyaan,Fraud Saiyyan,0,2019,109,Comedy|Drama,4.2,504,Fraud Saiyyan is the story of a con artist in ...,Fraud Saiyyan is the story of a con artist in ...,NaN,Arshad Warsi|Saurabh Shukla|Flora Saini|Sara L...,NaN
7,Bombairiya,tt4971258,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Bombairiya,Bombairiya,Bombairiya,0,2019,104,Comedy|Crime|Drama,4.3,295,It follows the story of Meghna who gets embro...,It follows the story of Meghna who gets embro...,They didn't mean to change the world.,Radhika Apte|Akshay Oberoi|Siddhanth Kapoor|Ra...,NaN
8,Manikarnika: The Queen of Jhansi,tt6903440,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Manikarnika:_The...,Manikarnika: The Queen of Jhansi,Manikarnika: The Queen of Jhansi,0,2019,148,Action|Biography|Drama,6.5,7361,Manikarnika born in Varanasi when Dixt a minis...,Story of Rani Lakshmibai one of the leading f...,NaN,Kangana Ranaut|Rimi Sen|Atul Kulkarni|Nalneesh...,NaN
9,Thackeray (film),tt7777196,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Thackeray_(film),Thackeray,Thackeray,0,2019,120,Biography|Drama,5.1,2301,Balasaheb Thackrey works as a cartoonist for a...,Biographical account of Shiv Sena Supremo Bal...,NaN,Nawazuddin Siddiqui|Amrita Rao|Abdul Quadir Am...,NaN


In [92]:
stock_data_modified

,Quantity,Price per share,Total Amount
Stock Name,,,
Eternal,10,230,2300
Jio Financial,10,230,2300


In [93]:
stock_data_modified.loc[['Eternal'],['Quantity','Total Amount']] # extract only eternal's quantity and amount invested

,Quantity,Total Amount
Stock Name,,
Eternal,10,2300


In [95]:
stock_data_modified.iloc[0,0:] # extract 0th row 

Quantity             10
Price per share     230
Total Amount       2300
Name: Eternal, dtype: int64

#### fancy indexing is also allowed in .iloc[ ] and .loc[ ]

### Filtering the DataFrame

##### Find all the final winners

In [97]:
ipl_data.head(1)

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
0,1312200,Ahmedabad,2022-05-29,2022,Final,Rajasthan Royals,Gujarat Titans,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,bat,N,Gujarat Titans,Wickets,7.0,NaN,HH Pandya,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",CB Gaffaney,Nitin Menon


In [112]:
criteria = ipl_data.MatchNumber == 'Final'
criteria

0       True
1      False
2      False
3      False
4      False
       ...  
945    False
946    False
947    False
948    False
949    False
Name: MatchNumber, Length: 950, dtype: bool

In [115]:
ipl_data[criteria]

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
0,1312200,Ahmedabad,2022-05-29,2022,Final,Rajasthan Royals,Gujarat Titans,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,bat,N,Gujarat Titans,Wickets,7.0,NaN,HH Pandya,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",CB Gaffaney,Nitin Menon
74,1254117,Dubai,2021-10-15,2021,Final,Chennai Super Kings,Kolkata Knight Riders,Dubai International Cricket Stadium,Kolkata Knight Riders,field,N,Chennai Super Kings,Runs,27.0,NaN,F du Plessis,"['RD Gaikwad', 'F du Plessis', 'RV Uthappa', '...","['Shubman Gill', 'VR Iyer', 'N Rana', 'SP Nari...",Nitin Menon,RK Illingworth
134,1237181,NaN,2020-11-10,2020/21,Final,Delhi Capitals,Mumbai Indians,Dubai International Cricket Stadium,Delhi Capitals,bat,N,Mumbai Indians,Wickets,5.0,NaN,TA Boult,"['MP Stoinis', 'S Dhawan', 'AM Rahane', 'SS Iy...","['RG Sharma', 'Q de Kock', 'SA Yadav', 'Ishan ...",CB Gaffaney,Nitin Menon
194,1181768,Hyderabad,2019-05-12,2019,Final,Mumbai Indians,Chennai Super Kings,Rajiv Gandhi International Stadium,Mumbai Indians,bat,N,Mumbai Indians,Runs,1.0,NaN,JJ Bumrah,"['Q de Kock', 'RG Sharma', 'SA Yadav', 'Ishan ...","['F du Plessis', 'SR Watson', 'SK Raina', 'AT ...",IJ Gould,Nitin Menon
254,1136620,Mumbai,2018-05-27,2018,Final,Sunrisers Hyderabad,Chennai Super Kings,Wankhede Stadium,Chennai Super Kings,field,N,Chennai Super Kings,Wickets,8.0,NaN,SR Watson,"['SP Goswami', 'S Dhawan', 'KS Williamson', 'S...","['SR Watson', 'F du Plessis', 'SK Raina', 'AT ...",M Erasmus,S Ravi
314,1082650,Hyderabad,2017-05-21,2017,Final,Mumbai Indians,Rising Pune Supergiant,"Rajiv Gandhi International Stadium, Uppal",Mumbai Indians,bat,N,Mumbai Indians,Runs,1.0,NaN,KH Pandya,"['LMP Simmons', 'PA Patel', 'AT Rayudu', 'RG S...","['AM Rahane', 'RA Tripathi', 'SPD Smith', 'MS ...",NJ Llong,S Ravi
373,981019,Bangalore,2016-05-29,2016,Final,Royal Challengers Bangalore,Sunrisers Hyderabad,M Chinnaswamy Stadium,Sunrisers Hyderabad,bat,N,Sunrisers Hyderabad,Runs,8.0,NaN,BCJ Cutting,"['CH Gayle', 'V Kohli', 'AB de Villiers', 'KL ...","['DA Warner', 'S Dhawan', 'MC Henriques', 'Yuv...",HDPK Dharmasena,BNJ Oxenford
433,829823,Kolkata,2015-05-24,2015,Final,Mumbai Indians,Chennai Super Kings,Eden Gardens,Chennai Super Kings,field,N,Mumbai Indians,Runs,41.0,NaN,RG Sharma,"['LMP Simmons', 'PA Patel', 'RG Sharma', 'KA P...","['DR Smith', 'MEK Hussey', 'SK Raina', 'MS Dho...",HDPK Dharmasena,RK Illingworth
492,734049,Bangalore,2014-06-01,2014,Final,Kolkata Knight Riders,Kings XI Punjab,M Chinnaswamy Stadium,Kolkata Knight Riders,field,N,Kolkata Knight Riders,Wickets,3.0,NaN,MK Pandey,"['RV Uthappa', 'G Gambhir', 'MK Pandey', 'YK P...","['V Sehwag', 'M Vohra', 'GJ Bailey', 'WP Saha'...",HDPK Dharmasena,BNJ Oxenford
552,598073,Kolkata,2013-05-26,2013,Final,Chennai Super Kings,Mumbai Indians,Eden Gardens,Mumbai Indians,bat,N,Mumbai Indians,Runs,23.0,NaN,KA Pollard,"['MEK Hussey', 'M Vijay', 'SK Raina', 'S Badri...","['DR Smith', 'AP Tare', 'KD Karthik', 'RG Shar...",HDPK Dharmasena,SJA Taufel


In [114]:
ipl_data[criteria].loc[:,['Season','WinningTeam']]

,Season,WinningTeam
0,2022,Gujarat Titans
74,2021,Chennai Super Kings
134,2020/21,Mumbai Indians
194,2019,Mumbai Indians
254,2018,Chennai Super Kings
314,2017,Mumbai Indians
373,2016,Sunrisers Hyderabad
433,2015,Mumbai Indians
492,2014,Kolkata Knight Riders
552,2013,Mumbai Indians


#### How many super over finishes have occured

In [118]:
ipl_data.sample()

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
523,733985,Delhi,2014-05-07,2014,28,Delhi Daredevils,Kolkata Knight Riders,Feroz Shah Kotla,Delhi Daredevils,bat,N,Kolkata Knight Riders,Wickets,8.0,NaN,G Gambhir,"['Q de Kock', 'M Vijay', 'KP Pietersen', 'KD K...","['RV Uthappa', 'G Gambhir', 'MK Pandey', 'JH K...",BNJ Oxenford,C Shamshuddin


In [122]:
ipl_data['SuperOver'].unique()

array(['N', 'Y', nan], dtype=object)

In [130]:
criteria = ipl_data['SuperOver'] == 'Y'
criteria.value_counts() #True means it has super overs

SuperOver
False    936
True      14
Name: count, dtype: int64

In [132]:
criteria.value_counts()[True]

np.int64(14)

#### how many matches has csk won in kolkata

In [134]:
ipl_data.sample(10)

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
879,392194,Port Elizabeth,2009-04-26,2009,14,Royal Challengers Bangalore,Delhi Daredevils,St George's Park,Royal Challengers Bangalore,bat,N,Delhi Daredevils,Wickets,6.0,NaN,TM Dilshan,"['JH Kallis', 'RV Uthappa', 'KP Pietersen', 'L...","['G Gambhir', 'V Sehwag', 'TM Dilshan', 'AB de...",S Asnani,BG Jerling
200,1178429,Bengaluru,2019-05-04,2019,54,Sunrisers Hyderabad,Royal Challengers Bangalore,M.Chinnaswamy Stadium,Royal Challengers Bangalore,field,N,Royal Challengers Bangalore,Wickets,4.0,NaN,SO Hetmyer,"['WP Saha', 'MJ Guptill', 'MK Pandey', 'KS Wil...","['PA Patel', 'V Kohli', 'AB de Villiers', 'SO ...",AK Chaudhary,NJ Llong
147,1216524,NaN,2020-10-27,2020/21,47,Sunrisers Hyderabad,Delhi Capitals,Dubai International Cricket Stadium,Delhi Capitals,field,N,Sunrisers Hyderabad,Runs,88.0,NaN,WP Saha,"['DA Warner', 'WP Saha', 'MK Pandey', 'KS Will...","['AM Rahane', 'S Dhawan', 'MP Stoinis', 'SO He...",AK Chaudhary,Nitin Menon
116,1254075,Mumbai,2021-04-24,2021,18,Kolkata Knight Riders,Rajasthan Royals,"Wankhede Stadium, Mumbai",Rajasthan Royals,field,N,Rajasthan Royals,Wickets,6.0,NaN,CH Morris,"['N Rana', 'Shubman Gill', 'RA Tripathi', 'SP ...","['JC Buttler', 'YBK Jaiswal', 'SV Samson', 'S ...",Navdeep Singh,S Ravi
744,501228,Mumbai,2011-04-27,2011,31,Pune Warriors,Chennai Super Kings,Dr DY Patil Sports Academy,Pune Warriors,bat,N,Chennai Super Kings,Wickets,8.0,NaN,DE Bollinger,"['JD Ryder', 'MD Mishra', 'MK Pandey', 'M Manh...","['MEK Hussey', 'M Vijay', 'S Badrinath', 'SK R...",Asad Rauf,SL Shastri
169,1216523,Abu Dhabi,2020-10-10,2020/21,24,Kolkata Knight Riders,Kings XI Punjab,Sheikh Zayed Stadium,Kolkata Knight Riders,bat,N,Kolkata Knight Riders,Runs,2.0,NaN,KD Karthik,"['RA Tripathi', 'Shubman Gill', 'N Rana', 'EJG...","['KL Rahul', 'MA Agarwal', 'N Pooran', 'P Simr...",UV Gandhe,CB Gaffaney
386,980991,Chandigarh,2016-05-15,2016,46,Kings XI Punjab,Sunrisers Hyderabad,"Punjab Cricket Association IS Bindra Stadium, ...",Kings XI Punjab,bat,N,Sunrisers Hyderabad,Wickets,7.0,NaN,HM Amla,"['HM Amla', 'M Vijay', 'WP Saha', 'Gurkeerat S...","['DA Warner', 'S Dhawan', 'DJ Hooda', 'Yuvraj ...",KN Ananthapadmanabhan,M Erasmus
461,829769,Mumbai,2015-05-01,2015,32,Mumbai Indians,Rajasthan Royals,Wankhede Stadium,Rajasthan Royals,field,N,Mumbai Indians,Runs,8.0,NaN,AT Rayudu,"['LMP Simmons', 'PA Patel', 'UBT Chand', 'RG S...","['AM Rahane', 'SR Watson', 'SV Samson', 'SPD S...",HDPK Dharmasena,CK Nandan
858,392217,Kimberley,2009-05-09,2009,37,Chennai Super Kings,Rajasthan Royals,De Beers Diamond Oval,Rajasthan Royals,bat,N,Chennai Super Kings,Wickets,7.0,NaN,S Badrinath,"['M Vijay', 'ML Hayden', 'SK Raina', 'S Badrin...","['GC Smith', 'NV Ojha', 'SA Asnodkar', 'RA Jad...",GAV Baxter,HDPK Dharmasena
341,1082622,Kolkata,2017-04-28,2017,32,Kolkata Knight Riders,Delhi Daredevils,Eden Gardens,Kolkata Knight Riders,field,N,Kolkata Knight Riders,Wickets,7.0,NaN,G Gambhir,"['SP Narine', 'G Gambhir', 'RV Uthappa', 'MK P...","['SV Samson', 'KK Nair', 'SS Iyer', 'RR Pant',...",NJ Llong,S Ravi


In [136]:
criteria = (ipl_data['WinningTeam'] == 'Chennai Super Kings') & (ipl_data['City'] == 'Kolkata')
criteria

0      False
1      False
2      False
3      False
4      False
       ...  
945    False
946    False
947    False
948    False
949    False
Length: 950, dtype: bool

In [137]:
ipl_data[criteria]

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
224,1178404,Kolkata,2019-04-14,2019,29,Kolkata Knight Riders,Chennai Super Kings,Eden Gardens,Chennai Super Kings,field,N,Chennai Super Kings,Wickets,5.0,NaN,Imran Tahir,"['CA Lynn', 'SP Narine', 'N Rana', 'RV Uthappa...","['SR Watson', 'F du Plessis', 'SK Raina', 'AT ...",CK Nandan,RJ Tucker
602,598022,Kolkata,2013-04-20,2013,26,Kolkata Knight Riders,Chennai Super Kings,Eden Gardens,Kolkata Knight Riders,bat,N,Chennai Super Kings,Wickets,4.0,NaN,RA Jadeja,"['G Gambhir', 'YK Pathan', 'JH Kallis', 'EJG M...","['MEK Hussey', 'R Ashwin', 'M Vijay', 'SK Rain...",Asad Rauf,AK Chaudhary
641,548368,Kolkata,2012-05-14,2012,63,Kolkata Knight Riders,Chennai Super Kings,Eden Gardens,Chennai Super Kings,field,N,Chennai Super Kings,Wickets,5.0,NaN,MEK Hussey,"['G Gambhir', 'BB McCullum', 'JH Kallis', 'MK ...","['MEK Hussey', 'M Vijay', 'SK Raina', 'F du Pl...",JD Cloete,SJA Taufel
827,419113,Kolkata,2010-03-16,2009/10,8,Kolkata Knight Riders,Chennai Super Kings,Eden Gardens,Chennai Super Kings,bat,N,Chennai Super Kings,Runs,55.0,NaN,MS Dhoni,"['BJ Hodge', 'MK Tiwary', 'SC Ganguly', 'WP Sa...","['M Vijay', 'ML Hayden', 'SK Raina', 'S Badrin...",HDPK Dharmasena,AM Saheba
908,336025,Kolkata,2008-05-18,2007/08,41,Kolkata Knight Riders,Chennai Super Kings,Eden Gardens,Kolkata Knight Riders,bat,N,Chennai Super Kings,Runs,3.0,D/L,M Ntini,"['Salman Butt', 'Mohammad Hafeez', 'SC Ganguly...","['PA Patel', 'SP Fleming', 'S Vidyut', 'SK Rai...",Asad Rauf,K Hariharan


In [140]:
ipl_data[criteria].shape #row respresents the count

(5, 20)

In [141]:
ipl_data[criteria].shape[0]

5

#### toss winner is the match winner in percentage 

In [144]:
ipl_data.sample()

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
688,548319,Chandigarh,2012-04-12,2012,14,Kings XI Punjab,Pune Warriors,"Punjab Cricket Association Stadium, Mohali",Kings XI Punjab,field,N,Kings XI Punjab,Wickets,7.0,NaN,AD Mascarenhas,"['PC Valthaty', 'AC Gilchrist', 'SE Marsh', 'M...","['JD Ryder', 'SC Ganguly', 'MN Samuels', 'RV U...",VA Kulkarni,SK Tarapore


In [145]:
criteria = ipl_data['TossWinner'] == ipl_data['WinningTeam']
criteria

0      False
1       True
2      False
3       True
4      False
       ...  
945    False
946    False
947    False
948     True
949    False
Length: 950, dtype: bool

In [146]:
ipl_data[criteria]

,ID,City,Date,Season,MatchNumber,Team1,Team2,Venue,TossWinner,TossDecision,SuperOver,WinningTeam,WonBy,Margin,method,Player_of_Match,Team1Players,Team2Players,Umpire1,Umpire2
1,1312199,Ahmedabad,2022-05-27,2022,Qualifier 2,Royal Challengers Bangalore,Rajasthan Royals,"Narendra Modi Stadium, Ahmedabad",Rajasthan Royals,field,N,Rajasthan Royals,Wickets,7.0,NaN,JC Buttler,"['V Kohli', 'F du Plessis', 'RM Patidar', 'GJ ...","['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...",CB Gaffaney,Nitin Menon
3,1312197,Kolkata,2022-05-24,2022,Qualifier 1,Rajasthan Royals,Gujarat Titans,"Eden Gardens, Kolkata",Gujarat Titans,field,N,Gujarat Titans,Wickets,7.0,NaN,DA Miller,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['WP Saha', 'Shubman Gill', 'MS Wade', 'HH Pan...",BNJ Oxenford,VK Sharma
5,1304115,Mumbai,2022-05-21,2022,69,Delhi Capitals,Mumbai Indians,"Wankhede Stadium, Mumbai",Mumbai Indians,field,N,Mumbai Indians,Wickets,5.0,NaN,JJ Bumrah,"['PP Shaw', 'DA Warner', 'MR Marsh', 'RR Pant'...","['Ishan Kishan', 'RG Sharma', 'D Brevis', 'Til...",Nitin Menon,Tapan Sharma
8,1304112,Navi Mumbai,2022-05-18,2022,66,Lucknow Super Giants,Kolkata Knight Riders,"Dr DY Patil Sports Academy, Mumbai",Lucknow Super Giants,bat,N,Lucknow Super Giants,Runs,2.0,NaN,Q de Kock,"['Q de Kock', 'KL Rahul', 'E Lewis', 'DJ Hooda...","['VR Iyer', 'A Tomar', 'N Rana', 'SS Iyer', 'S...",R Pandit,YC Barde
11,1304109,Mumbai,2022-05-15,2022,63,Rajasthan Royals,Lucknow Super Giants,"Brabourne Stadium, Mumbai",Rajasthan Royals,bat,N,Rajasthan Royals,Runs,24.0,NaN,TA Boult,"['YBK Jaiswal', 'JC Buttler', 'SV Samson', 'D ...","['Q de Kock', 'KL Rahul', 'A Badoni', 'DJ Hood...",PG Pathak,Tapan Sharma
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
935,335996,Bangalore,2008-04-28,2007/08,15,Royal Challengers Bangalore,Chennai Super Kings,M Chinnaswamy Stadium,Chennai Super Kings,bat,N,Chennai Super Kings,Runs,13.0,NaN,MS Dhoni,"['B Chipli', 'W Jaffer', 'LRPL Taylor', 'JH Ka...","['PA Patel', 'ML Hayden', 'MEK Hussey', 'SK Ra...",BR Doctrove,RB Tiffin
937,335994,Mumbai,2008-04-27,2007/08,14,Mumbai Indians,Deccan Chargers,Dr DY Patil Sports Academy,Deccan Chargers,field,N,Deccan Chargers,Wickets,10.0,NaN,AC Gilchrist,"['L Ronchi', 'ST Jayasuriya', 'AM Rahane', 'RV...","['AC Gilchrist', 'VVS Laxman', 'A Symonds', 'R...",Asad Rauf,SL Shastri
939,335992,Bangalore,2008-04-26,2007/08,12,Royal Challengers Bangalore,Rajasthan Royals,M Chinnaswamy Stadium,Rajasthan Royals,field,N,Rajasthan Royals,Wickets,7.0,NaN,SR Watson,"['S Chanderpaul', 'R Dravid', 'LRPL Taylor', '...","['GC Smith', 'M Rawat', 'YK Pathan', 'SR Watso...",MR Benson,IL Howell
941,335990,Hyderabad,2008-04-24,2007/08,9,Deccan Chargers,Rajasthan Royals,"Rajiv Gandhi International Stadium, Uppal",Rajasthan Royals,field,N,Rajasthan Royals,Wickets,3.0,NaN,YK Pathan,"['AC Gilchrist', 'VVS Laxman', 'Shahid Afridi'...","['Kamran Akmal', 'GC Smith', 'YK Pathan', 'SR ...",Asad Rauf,MR Benson


In [151]:
ipl_data[criteria].shape #rows represents how many times a team wins both toss and game

(489, 20)

In [ ]:
ipl_data.shape # rows represents total game

(950, 20)

In [153]:
# percentage

((ipl_data[criteria].shape[0]) / (ipl_data.shape[0])) * 100

51.473684210526315

#### movies with rating higher than 8 and votes>10000

In [155]:
movies_data.sample()

,title_x,imdb_id,poster_path,wiki_link,title_y,original_title,is_adult,year_of_release,runtime,genres,imdb_rating,imdb_votes,story,summary,tagline,actors,wins_nominations,release_date
870,Once Upon a Time in Mumbaai,tt1395054,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Once_Upon_a_Time...,Once Upon a Time in Mumbai,Once Upon a Time in Mumbaai,0,2010,134,Action|Crime|Drama,7.4,15169,During March 1993 after his Ambassador car is...,A smuggler rises to power in in 1970s Mumbai ...,The man Mumbai once feared. The man Mumbai now...,Ajay Devgn|Emraan Hashmi|Kangana Ranaut|Prachi...,7 wins & 33 nominations,30 July 2010 (India)


In [156]:
criteria = (movies_data['imdb_rating'] > 8) & (movies_data['imdb_votes'] > 10000)
criteria

0        True
1       False
2       False
3       False
4       False
        ...  
1624    False
1625    False
1626    False
1627    False
1628    False
Length: 1629, dtype: bool

In [158]:
movies_data[criteria].loc[:, 'original_title']

0                Uri: The Surgical Strike
11                              Gully Boy
37                             Article 15
40                               Super 30
143                               Tumbbad
146                             Andhadhun
325                                  Pink
354                                Dangal
418                                Masaan
426                              Drishyam
436                                Talvar
469                                 Queen
536                                Haider
566                                  Ugly
567                                    PK
589                          Vishwaroopam
612                    Bhaag Milkha Bhaag
638                                Shahid
668                      Paan Singh Tomar
669                               Kahaani
693                    Gangs of Wasseypur
694                    Gangs of Wasseypur
709                                Barfi!
714                       OMG: Oh 

#### Action movies with rating higher than 7.5

In [165]:
movies_data.head(5)

,title_x,imdb_id,poster_path,wiki_link,title_y,original_title,is_adult,year_of_release,runtime,genres,imdb_rating,imdb_votes,story,summary,tagline,actors,wins_nominations,release_date
0,Uri: The Surgical Strike,tt8291224,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Uri:_The_Surgica...,Uri: The Surgical Strike,Uri: The Surgical Strike,0,2019,138,Action|Drama|War,8.4,35112,Divided over five chapters the film chronicle...,Indian army special forces execute a covert op...,NaN,Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...,4 wins,11 January 2019 (USA)
1,Battalion 609,tt9472208,NaN,https://en.wikipedia.org/wiki/Battalion_609,Battalion 609,Battalion 609,0,2019,131,War,4.1,73,The story revolves around a cricket match betw...,The story of Battalion 609 revolves around a c...,NaN,Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...,NaN,11 January 2019 (India)
2,The Accidental Prime Minister (film),tt6986710,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/The_Accidental_P...,The Accidental Prime Minister,The Accidental Prime Minister,0,2019,112,Biography|Drama,6.1,5549,Based on the memoir by Indian policy analyst S...,Explores Manmohan Singh's tenure as the Prime ...,NaN,Anupam Kher|Akshaye Khanna|Aahana Kumra|Atul S...,NaN,11 January 2019 (USA)
3,Why Cheat India,tt8108208,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Why_Cheat_India,Why Cheat India,Why Cheat India,0,2019,121,Crime|Drama,6.0,1891,The movie focuses on existing malpractices in ...,The movie focuses on existing malpractices in ...,NaN,Emraan Hashmi|Shreya Dhanwanthary|Snighdadeep ...,NaN,18 January 2019 (USA)
4,Evening Shadows,tt6028796,NaN,https://en.wikipedia.org/wiki/Evening_Shadows,Evening Shadows,Evening Shadows,0,2018,102,Drama,7.3,280,While gay rights and marriage equality has bee...,Under the 'Evening Shadows' truth often plays...,NaN,Mona Ambegaonkar|Ananth Narayan Mahadevan|Deva...,17 wins & 1 nomination,11 January 2019 (India)


In [172]:
criteria1 = movies_data['genres'].apply(lambda x: True if 'Action' in x else False) # to get action genres movie

criteria2 = movies_data['imdb_rating'] > 7.5

In [173]:
criteria1

0        True
1       False
2       False
3       False
4       False
        ...  
1624    False
1625    False
1626    False
1627     True
1628    False
Name: genres, Length: 1629, dtype: bool

In [174]:
criteria2

0        True
1       False
2       False
3       False
4       False
        ...  
1624    False
1625    False
1626    False
1627    False
1628     True
Name: imdb_rating, Length: 1629, dtype: bool

In [175]:
criteria = criteria1 & criteria2

In [176]:
criteria

0        True
1       False
2       False
3       False
4       False
        ...  
1624    False
1625    False
1626    False
1627    False
1628    False
Length: 1629, dtype: bool

In [179]:
movies_data[criteria].loc[:,'original_title']

0            Uri: The Surgical Strike
41               Family of Thakurganj
84                          Mukkabaaz
106                             Raazi
110     Parmanu: The Story of Pokhran
112           Bhavesh Joshi Superhero
169                  The Ghazi Attack
219                         Raag Desh
258                     Saala Khadoos
280                         Laal Rang
297                       Udta Punjab
354                            Dangal
362                 Bajrangi Bhaijaan
365                              Baby
393        Detective Byomkesh Bakshy!
449                             Titli
536                            Haider
589                      Vishwaroopam
625                       Madras Cafe
668                  Paan Singh Tomar
693                Gangs of Wasseypur
694                Gangs of Wasseypur
982                      Jodhaa Akbar
1039                             1971
1058                     Black Friday
1188                           Omkara
1293        

In [181]:
movies_data[criteria].shape[0] # total criteria satisfied movie

33

### Write a function that can return the track record of 2 teams against each other

In [184]:
team1 = 'Kings XI Punjab'
team2 = 'Royal Challengers Bangalore'

In [187]:
criteria1 = ipl_data['Team1'].isin([team1, team2])
criteria2 = ipl_data['Team2'].isin([team1, team2])

criteria = criteria1 & criteria2
criteria

0      False
1      False
2      False
3      False
4      False
       ...  
945    False
946    False
947    False
948    False
949    False
Length: 950, dtype: bool

In [203]:
winning_matches = ipl_data[criteria].WinningTeam.value_counts()[0]
winning_matches

/var/folders/94/2_5s7zqj0cqd50nz6b9khlxh0000gp/T/ipykernel_19389/3861785846.py:1: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  winning_matches = ipl_data[criteria].WinningTeam.value_counts()[0]


np.int64(14)

In [202]:
winning_matches.index[0]

'Kings XI Punjab'

In [213]:
def track_record(team1, team2):
    criteria1 = ipl_data['Team1'].isin([team1,team2])
    criteria2 = ipl_data['Team2'].isin([team1,team2])

    criteria = criteria1 & criteria2

    t1_name = ipl_data[criteria].WinningTeam.value_counts().index[0]
    t1_winning_game = ipl_data[criteria].WinningTeam.value_counts()[0]

    t2_name = ipl_data[criteria].WinningTeam.value_counts().index[1]
    t2_winning_game = ipl_data[criteria].WinningTeam.value_counts()[1]

    t1_winning_ratio = (t1_winning_game/(t1_winning_game+t2_winning_game))*100
    t2_winning_ratio = (t2_winning_game/(t1_winning_game+t2_winning_game))*100

    print('Team', t1_name, 'has the win ratio of', t1_winning_ratio, 'against', t2_name)
    print('Team', t2_name, 'has the win ratio of', t2_winning_ratio, 'against', t1_name)

    # return t1_name,t2_name,t1_winning_ratio,t2_winning_ratio



In [214]:
team1 = 'Kings XI Punjab'
team2 = 'Royal Challengers Bangalore'
track_record(team1, team2)

Team Kings XI Punjab has the win ratio of 53.84615384615385 against Royal Challengers Bangalore
Team Royal Challengers Bangalore has the win ratio of 46.15384615384615 against Kings XI Punjab


/var/folders/94/2_5s7zqj0cqd50nz6b9khlxh0000gp/T/ipykernel_19389/3393824468.py:8: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  t1_winning_game = ipl_data[criteria].WinningTeam.value_counts()[0]
/var/folders/94/2_5s7zqj0cqd50nz6b9khlxh0000gp/T/ipykernel_19389/3393824468.py:11: FutureWarning: Series.__getitem__ treating keys as positions is deprecated. In a future version, integer keys will always be treated as labels (consistent with DataFrame behavior). To access a value by position, use `ser.iloc[pos]`
  t2_winning_game = ipl_data[criteria].WinningTeam.value_counts()[1]


### Creating new column

In [242]:
movies_data['Country'] = 'India'
movies_data

,title_x,imdb_id,poster_path,wiki_link,title_y,original_title,is_adult,year_of_release,runtime,genres,imdb_rating,imdb_votes,story,summary,tagline,actors,wins_nominations,release_date,Country
0,Uri: The Surgical Strike,tt8291224,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Uri:_The_Surgica...,Uri: The Surgical Strike,Uri: The Surgical Strike,0,2019,138,Action|Drama|War,8.4,35112,Divided over five chapters the film chronicle...,Indian army special forces execute a covert op...,NaN,Vicky Kaushal|Paresh Rawal|Mohit Raina|Yami Ga...,4 wins,11 January 2019 (USA),India
1,Battalion 609,tt9472208,NaN,https://en.wikipedia.org/wiki/Battalion_609,Battalion 609,Battalion 609,0,2019,131,War,4.1,73,The story revolves around a cricket match betw...,The story of Battalion 609 revolves around a c...,NaN,Vicky Ahuja|Shoaib Ibrahim|Shrikant Kamat|Elen...,NaN,11 January 2019 (India),India
2,The Accidental Prime Minister (film),tt6986710,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/The_Accidental_P...,The Accidental Prime Minister,The Accidental Prime Minister,0,2019,112,Biography|Drama,6.1,5549,Based on the memoir by Indian policy analyst S...,Explores Manmohan Singh's tenure as the Prime ...,NaN,Anupam Kher|Akshaye Khanna|Aahana Kumra|Atul S...,NaN,11 January 2019 (USA),India
3,Why Cheat India,tt8108208,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Why_Cheat_India,Why Cheat India,Why Cheat India,0,2019,121,Crime|Drama,6.0,1891,The movie focuses on existing malpractices in ...,The movie focuses on existing malpractices in ...,NaN,Emraan Hashmi|Shreya Dhanwanthary|Snighdadeep ...,NaN,18 January 2019 (USA),India
4,Evening Shadows,tt6028796,NaN,https://en.wikipedia.org/wiki/Evening_Shadows,Evening Shadows,Evening Shadows,0,2018,102,Drama,7.3,280,While gay rights and marriage equality has bee...,Under the 'Evening Shadows' truth often plays...,NaN,Mona Ambegaonkar|Ananth Narayan Mahadevan|Deva...,17 wins & 1 nomination,11 January 2019 (India),India
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1624,Tera Mera Saath Rahen,tt0301250,https://upload.wikimedia.org/wikipedia/en/2/2b...,https://en.wikipedia.org/wiki/Tera_Mera_Saath_...,Tera Mera Saath Rahen,Tera Mera Saath Rahen,0,2001,148,Drama,4.9,278,Raj Dixit lives with his younger brother Rahu...,A man is torn between his handicapped brother ...,NaN,Ajay Devgn|Sonali Bendre|Namrata Shirodkar|Pre...,NaN,7 November 2001 (India),India
1625,Yeh Zindagi Ka Safar,tt0298607,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Yeh_Zindagi_Ka_S...,Yeh Zindagi Ka Safar,Yeh Zindagi Ka Safar,0,2001,146,Drama,3.0,133,Hindi pop-star Sarina Devan lives a wealthy ...,A singer finds out she was adopted when the ed...,NaN,Ameesha Patel|Jimmy Sheirgill|Nafisa Ali|Gulsh...,NaN,16 November 2001 (India),India
1626,Sabse Bada Sukh,tt0069204,NaN,https://en.wikipedia.org/wiki/Sabse_Bada_Sukh,Sabse Bada Sukh,Sabse Bada Sukh,0,2018,\N,Comedy|Drama,6.1,13,Village born Lalloo re-locates to Bombay and ...,Village born Lalloo re-locates to Bombay and ...,NaN,Vijay Arora|Asrani|Rajni Bala|Kumud Damle|Utpa...,NaN,NaN,India
1627,Daaka,tt10833860,https://upload.wikimedia.org/wikipedia/en/thum...,https://en.wikipedia.org/wiki/Daaka,Daaka,Daaka,0,2019,136,Action,7.4,38,Shinda tries robbing a bank so he can be wealt...,Shinda tries robbing a bank so he can be wealt...,NaN,Gippy Grewal|Zareen Khan|,NaN,1 November 2019 (USA),India
